File: java/src/main/be.steinsomers.bron_kerbosch/BronKerbosch3_ST.java
package be.steinsomers.bron_kerbosch;

import java.util.HashSet;
import java.util.Set;
import java.util.stream.Collectors;

public final class BronKerbosch3_ST implements BronKerboschAlgorithm {
    @Override
    public void explore(final UndirectedGraph graph, final CliqueConsumer cliqueConsumer) {
        final var worker = new Worker(graph, cliqueConsumer);
        worker.work();
    }

    private record VisitJob(int startVertex, Set<Integer> candidates, Set<Integer> excluded) {
    }

    private record Worker(UndirectedGraph graph, CliqueConsumer cliqueConsumer) {
        void work() {
            final var visitProducer = new VisitProducer();
            final var visitor = new Visitor();
            final var ordering = new DegeneracyIterator(graph);
            ordering.stream()
                    .mapToObj(visitProducer::createJob)
                    .toList()
                    .parallelStream()
                    .forEach(visitor::visit);
        }

        private final class VisitProducer {
            private final boolean[] excluded = new boolean[graph.order()];

            VisitJob createJob(final Integer startVtx) {
                final var neighbours = graph.neighbours(startVtx);
                final var neighbouringCandidates = neighbours.stream().filter(v -> !excluded[v])
                        .collect(Collectors.toCollection(HashSet::new));
                final var neighbouringExcluded = neighbours.stream().filter(v -> excluded[v])
                        .collect(Collectors.toCollection(HashSet::new));
                assert !neighbouringCandidates.isEmpty();
                excluded[startVtx] = true;
                return new VisitJob(startVtx, neighbouringCandidates, neighbouringExcluded);
            }
        }

        private final class Visitor {
            void visit(final VisitJob job) {
                BronKerboschPivot.visit(graph, cliqueConsumer,
                        PivotChoice.MaxDegreeLocal,
                        job.candidates,
                        job.excluded,
                        new int[]{job.startVertex});
            }
        }
    }
}

File: kotlin/src/main/be.steinsomers.bron_kerbosch/BronKerbosch3ST.kt
package be.steinsomers.bron_kerbosch

class BronKerbosch3ST : BronKerboschAlgorithm {
    override fun explore(graph: UndirectedGraph, cliqueConsumer: CliqueConsumer) {
        val worker = Worker(graph, cliqueConsumer)
        worker.work()
    }

    private sealed class VisitJob {
        // Similar to BronKerbosch3MT but doesn't need sentinels
        data class Work(
            val startVertex: Vertex,
            val candidates: MutableSet<Vertex>,
            val excluded: MutableSet<Vertex>
        ) : VisitJob()
    }

    private class Worker(private val graph: UndirectedGraph, private val cliqueConsumer: CliqueConsumer) {
        private val degeneracy = GraphDegeneracy(graph)

        fun work() {
            val visitProducer = VisitProducer()
            val visitor = Visitor()
            val storage = degeneracy.asSequence()
                .map { pick -> visitProducer.createJob(pick) }
                .toList() // TODO parallelize without intermediate list
                .parallelStream()
                .map(visitor::visit)
                .toList()
            cliqueConsumer.storage.absorb(storage)
        }

        private inner class VisitProducer {
            fun createJob(startVtx: Vertex): VisitJob {
                val (neighbouringCandidates, neighbouringExcluded) =
                    Util.partition(graph.neighbours(startVtx)) { v -> degeneracy.isCandidate(v) }
                Debug.assert { neighbouringCandidates.isNotEmpty() }
                return VisitJob.Work(startVtx, neighbouringCandidates, neighbouringExcluded)
            }
        }

        private inner class Visitor {
            fun visit(job: VisitJob): CliqueStorage {
                when (job) {
                    is VisitJob.Work -> {
                        val storage = cliqueConsumer.storage.spawn(1)[0]
                        BronKerboschPivot.visit(
                            graph = graph,
                            cliqueConsumer = CliqueConsumer(minSize = cliqueConsumer.minSize, storage = storage),
                            pivotChoice = PivotChoice.MaxDegreeLocal,
                            candidates = job.candidates,
                            excluded = job.excluded,
                            cliqueInProgress = Clique.singleton(job.startVertex)
                        )
                        return storage
                    }
                }
            }
        }
    }
}
